jonathanKingston/duckduckgo-privacy-extension-old
# First party fingerprinting improvements

**Reviewer:** @kdzwinel

## Description:

In this PR the change follows the design document submitted.

The changes are largely:
- Adding first party host specific fingerprinting generation.
- When canvas APIs are called they have randomized gradients applied per the first-party.
- When `navigator.plugins` is called the response is randomized and the entropy is reduced.

Changes to the design document are:
- To match the `agentSpoofer` the generation/lookup of the first party fingerprinting was moved to the background code in chrome-events.es6.js
- The storage mentioned was removed despite [being implemented](https://github.com/jonathanKingston/duckduckgo-privacy-extension/commit/1db924306a9d21390cc00e7e43af067a4d687eed)
  - The complexity is drastically increased by another storage just to minimise a small amount of breakage in websites that may be using a fingerprint as a persistent identifier.
  - The remaining work would be to have accessor observers in the content script to verify the fingerprint has been used and message back to the background script, this would then update the storage to ensure it's not cleared.
  - This additional code makes the storage lookup async which is a negative but also the additional complexity would make a lot of edge cases to test.
- The canvas gradient inner circle is centred in the larger circle to ensure that they are always inside each other. This prevents sometimes the generation of the gradient wasn't present (you can test this out by changing the transparency and calling the `toDataURI`)


## Steps to test this PR:
1. Go to https://panopticlick.eff.org/ without the extension changes in this PR
2. Compare the results with the extension loaded.

![image](https://user-images.githubusercontent.com/338988/97636195-a7bf0c00-1a30-11eb-9b1a-ecf6dbe05faa.png)

3. You should see both plugin and canvas "randomized by first party domain"
   - Firefox doesn't return a result for Plugins so that continues to have the same value.
   - As mentioned in the document WebGL gets the benefits of canvas however due to the shader and other properties aren't reduced here but could be fixed fairly quickly in a follow-up PR.

---

1. Open the developer console on a website you want to test:
```JavaScript
let script = document.createElement("script");
script.textContent = `
  async function initFingerprintJS() {
    const fp = await FingerprintJS.load();
    const result = await fp.get();
    console.log("Canvas fingerprint", result.components.canvas.value);
    console.log("Plugin fingerprint", result.components.plugins.value);
  }
`;
document.body.appendChild(script);
let script2 = document.createElement("script");
script2.src = "https://cdn.jsdelivr.net/npm/@fingerprintjs/fingerprintjs@3/dist/fp.min.js";
script2.addEventListener("load", initFingerprintJS);
document.body.appendChild(script2)
```
2. The results should be consistent for all frames in within a first-party for an hour.

## Automated tests:
  - One thing to note is the canvas checking in integration-test/background/test-fp-fingerprint.js specifically comparison to the frame sometimes returns a slightly different value. The canvas generation in both contexts are getting the same seed data; however, the response appears to race and I guess the gradient isn't applied uniformly. If intermittent failures become a problem this test could be disabled without removing the plugin code.

###### Reviewer Checklist:


###### PR Author Checklist:
